File: mobile/pwa/src/lib/device-identity.ts
/**
 * Device identity for the mobile PWA (docs/MOBILE-DESIGN.md §4.2).
 *
 * Ed25519 via @noble/curves (pure TS — avoids the WebCrypto support matrix).
 * The SPKI PEM and deviceId derivation MUST stay byte-compatible with the
 * desktop's src/main/remote/identity.ts (deviceIdFor hashes the PEM string);
 * scripts/test-pwa-pairing.mjs proves it end-to-end against a real host.
 */
import { ed25519 } from "@noble/curves/ed25519";
// v1.9.x: X25519 ships inside the ed25519 module (no standalone subpath export).
import { x25519 } from "@noble/curves/ed25519";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";

/** Fixed SPKI DER prefix for Ed25519 (SEQUENCE{SEQUENCE{OID 1.3.101.112}, BITSTRING}). */
const ED25519_SPKI_PREFIX = new Uint8Array([0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00]);

export interface DeviceIdentity {
  /** "device-<sha256(SPKI PEM) base64url[:24]>" — matches desktop deviceIdFor(). */
  deviceId: string;
  /** SPKI PEM of the Ed25519 public key (what the host verifies against). */
  publicKeyPem: string;
  /** X25519 public key (raw 32B, base64url) for E2E key agreement. */
  x25519PubB64u: string;
  /** X25519 private key — deterministically derived from the stored Ed25519 seed,
   * so exposing it adds no secret beyond what the keystore already holds. */
  x25519PrivB64u: string;
  /** Sign text → base64url signature (no padding), matching identity.signText. */
  signText(text: string): string;
}

/** Deterministic X25519 private key from the Ed25519 seed (single-seed device). */
export function deriveX25519Priv(ed25519Seed: Uint8Array): Uint8Array {
  return hkdf(sha256, ed25519Seed, new Uint8Array(0), "mpi-device-x25519", 32);
}

function bytesToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary);
}

export function toBase64Url(bytes: Uint8Array): string {
  return bytesToBase64(bytes).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function fromBase64Url(value: string): Uint8Array {
  const b64 = value.replace(/-/g, "+").replace(/_/g, "/");
  const padded = b64 + "=".repeat((4 - (b64.length % 4)) % 4);
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function utf8Bytes(text: string): Uint8Array {
  return new TextEncoder().encode(text);
}

/** Standard PEM with 64-char base64 lines — byte-identical to Node's spki/pem export. */
export function ed25519SpkiPem(publicKey32: Uint8Array): string {
  const der = new Uint8Array(ED25519_SPKI_PREFIX.length + publicKey32.length);
  der.set(ED25519_SPKI_PREFIX, 0);
  der.set(publicKey32, ED25519_SPKI_PREFIX.length);
  const b64 = bytesToBase64(der).match(/.{1,64}/g)?.join("\n") ?? "";
  return `-----BEGIN PUBLIC KEY-----\n${b64}\n-----END PUBLIC KEY-----\n`;
}

function deviceIdFor(publicKeyPem: string): string {
  return `device-${toBase64Url(sha256(utf8Bytes(publicKeyPem))).slice(0, 24)}`;
}

/** Create a fresh identity (or restore from a stored seed). */
export function createDeviceIdentity(seedB64url?: string): DeviceIdentity {
  const priv = seedB64url ? fromBase64Url(seedB64url) : ed25519.utils.randomPrivateKey();
  if (priv.length !== 32) throw new Error("invalid Ed25519 seed length");
  const publicKeyPem = ed25519SpkiPem(ed25519.getPublicKey(priv));
  const xPriv = deriveX25519Priv(priv);
  return {
    deviceId: deviceIdFor(publicKeyPem),
    publicKeyPem,
    x25519PubB64u: toBase64Url(x25519.getPublicKey(xPriv)),
    x25519PrivB64u: toBase64Url(xPriv),
    signText: (text) => toBase64Url(ed25519.sign(utf8Bytes(text), priv)),
  };
}

/** The storable private seed of a freshly generated identity. */
export function randomSeedB64url(): string {
  return toBase64Url(ed25519.utils.randomPrivateKey());
}

File: mobile/app/app/src/main/java/com/mpi/app/protocol/DeviceIdentity.kt
package com.mpi.app.protocol

import java.security.MessageDigest
import java.security.SecureRandom
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

/**
 * 设备身份 —— 与 PWA 的 mobile/pwa/src/lib/device-identity.ts 逐字节一致。
 *
 * ⚠️ 兼容性关键：`deviceId` 由 SPKI PEM 字符串的 SHA-256 得出，
 * 所以 PEM 的换行、padding、DER 前缀都必须与 TS 侧完全相同。
 */

/** 固定 SPKI DER 前缀（SEQUENCE{SEQUENCE{OID 1.3.101.112}, BITSTRING}）。 */
private val ED25519_SPKI_PREFIX = byteArrayOf(
    0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
)

private const val X25519_INFO = "mpi-device-x25519"

data class DeviceIdentity(
    /** "device-<sha256(SPKI PEM) base64url[:24]>" —— 与主机端 deviceIdFor() 一致。 */
    val deviceId: String,
    /** 公钥的 SPKI PEM（主机端据此验签）。 */
    val publicKeyPem: String,
    /** X25519 公钥（raw 32B，base64url），用于 E2E 密钥协商。 */
    val x25519PubB64u: String,
    /** X25519 私钥：由 Ed25519 种子确定性派生，不引入种子之外的新秘密。 */
    val x25519PrivB64u: String,
    /** 可持久化的 Ed25519 种子（base64url，32 字节）。 */
    val seedB64u: String,
) {
    /** text → base64url 签名（无 padding），与 identity.signText 一致。 */
    fun signText(text: String): String =
        Base64Url.encode(ed25519Sign(Base64Url.decode(seedB64u), text.toByteArray(Charsets.UTF_8)))
}

/** 由 Ed25519 种子确定性派生 X25519 私钥（单种子设备）。 */
fun deriveX25519Priv(ed25519Seed: ByteArray): ByteArray =
    Hkdf.sha256(ed25519Seed, ByteArray(0), X25519_INFO.toByteArray(Charsets.UTF_8), 32)

/** 标准 PEM（每行 64 字符）——与 Node 的 spki/pem 导出一致。 */
fun ed25519SpkiPem(publicKey32: ByteArray): String {
    val der = ED25519_SPKI_PREFIX + publicKey32
    val b64 = java.util.Base64.getEncoder().encodeToString(der)
    return "-----BEGIN PUBLIC KEY-----\n${b64.chunked(64).joinToString("\n")}\n-----END PUBLIC KEY-----\n"
}

fun deviceIdFor(publicKeyPem: String): String {
    val hash = MessageDigest.getInstance("SHA-256").digest(publicKeyPem.toByteArray(Charsets.UTF_8))
    return "device-" + Base64Url.encode(hash).take(24)
}

fun ed25519Sign(seed: ByteArray, message: ByteArray): ByteArray {
    val signer = Ed25519Signer()
    signer.init(true, Ed25519PrivateKeyParameters(seed, 0))
    signer.update(message, 0, message.size)
    return signer.generateSignature()
}

/** 新建身份；传入 seedB64u 则从已存种子恢复。 */
fun createDeviceIdentity(seedB64u: String? = null): DeviceIdentity {
    val seed = if (seedB64u != null) Base64Url.decode(seedB64u) else randomSeedBytes()
    require(seed.size == 32) { "invalid Ed25519 seed length" }

    val publicKey = Ed25519PrivateKeyParameters(seed, 0).generatePublicKey().encoded
    val pem = ed25519SpkiPem(publicKey)
    val x25519Priv = deriveX25519Priv(seed)
    return DeviceIdentity(
        deviceId = deviceIdFor(pem),
        publicKeyPem = pem,
        x25519PubB64u = Base64Url.encode(X25519.publicKeyFromPrivate(x25519Priv)),
        x25519PrivB64u = Base64Url.encode(x25519Priv),
        seedB64u = Base64Url.encode(seed),
    )
}

/** 全新身份的随机 Ed25519 种子（base64url，32 字节）。 */
fun randomSeedB64u(): String = Base64Url.encode(randomSeedBytes())

private fun randomSeedBytes(): ByteArray = ByteArray(32).also { SecureRandom().nextBytes(it) }
